Mixin contributing documentation (#1113)
ni/nimble: packages/nimble-components/CONTRIBUTING.md

});
```

> ### Leverage mixins for shared APIs across components

Reviewer: I think it would be really helpful if we could provide guidance of when to use one patter vs the other. Assuming we know that guidance, could you provide a few sentences giving that guidance?

Reviewer 2: +1, In particular, at least to start I'd recommend the Constrained Mixin pattern for any mixins we author in Nimble as it requires less manual type manipulation but continuing to use the FAST `applyMixin` helper for leveraging Mixins already defined in FastFoundation.

Typescript and the FAST library each offer patterns and/or mechanisms to alter the APIs for a component via a mixin.

> FAST provides an `applyMixins` function (which is just an implementation of the pattern described [here](https://www.typescriptlang.org/docs/handbook/mixins.html#alternative-pattern) in the Typscript docs) to alter the API of a given component with a set of provided mixin classes. This is used in several places within Nimble (see [here](https://github.com/ni/nimble/blob/6839ee05cf4d72efa6a20cd23e1d830047103745/packages/nimble-components/src/toggle-button/index.ts#L44) and [here](https://github.com/ni/nimble/blob/6839ee05cf4d72efa6a20cd23e1d830047103745/packages/nimble-components/src/anchor-tabs/index.ts#L274)).

Reviewer 2: Provide [descriptive link text](https://webaim.org/techniques/hypertext/link_text#uninformative) to improve readability of documentation generally for all readers and with the additional benefit of improved accessibility.

The way I train myself is that the link text itself should be descriptive enough standalone but shouldn't be a blue paragraph (just a couple words long), i.e:

> See how the [ToggleButton `StartEnd` mixin](#) is applied.

On its own it Is much easier to read documentation and gives me standalone context about what to expect when following the link. Also, you may remember that screen reader users had that shortcut to quickly navigate all links in a page, so multiple links with the link text `here` aren't helpful.

FAST provides an `applyMixins` function (which is just an implementation of the [Alternative Pattern](https://www.typescriptlang.org/docs/handbook/mixins.html#alternative-pattern) described in the Typscript docs) to alter the API of a given component with a set of provided mixin classes. For an example, see how the [ToggleButton StartEnd mixin](https://github.com/ni/nimble/blob/6839ee05cf4d72efa6a20cd23e1d830047103745/packages/nimble-components/src/toggle-button/index.ts#L44) is applied.

Another pattern in use within in Nimble is the [Constrained Mixin](https://www.typescriptlang.org/docs/handbook/mixins.html#constrained-mixins) pattern. An example in Nimble is the [FractionalWidth mixin](https://github.com/ni/nimble/blob/6839ee05cf4d72efa6a20cd23e1d830047103745/packages/nimble-components/src/table-column/mixins/fractional-width-column.ts#L16) which `TableColumnText`, for example, [ultimately extends](https://github.com/ni/nimble/blob/6839ee05cf4d72efa6a20cd23e1d830047103745/packages/nimble-components/src/table-column/text/index.ts#L61). This offers the ability for a mixin to extend the functionality of another concrete type and interface with its implementation.

> Our recommendation is to use the 'Constrained Mixin' pattern for applying mixins that are defined within Nimble, as they do not fundamentally alter existing types, but to use the `applyMixins` FAST method for consuming mixins available in the FAST library.

Reviewer: From everything I've seen, our docs are much more assertive than simply saying "Our recommendation is...". Perhaps a better way to say this is:

```suggestion
The 'Constrained Mixin' pattern is used for applying mixins that are defined within Nimble, as they do not fundamentally alter existing types, and the `applyMixins` FAST method is used for consuming mixins exported from the FAST library.
```
